### ISSUE TITLE
LSTM model compile time is very high

### ISSUE THREAD

**Issue by jramapuram** [fchollet/keras#133]
I understand the weights can be saved/loaded at a later time, however the time for an LSTM to compile is rather high.

If this can't be solved is there a solution by which we can serialize the model to prevent this initial overhead after the first generation? Prototyping LSTM's is currently take a pretty long time.

**Comment by patyork** [comment]
There is no way to serialize/save compiled Theano functions or models.

You can try turning off the graph optimizations for Theano, but I've found that this does not help much with recurrent neural networks; the automatic differentiation is where most of the compiling time comes from.

**Comment by jramapuram** [comment]
Maybe a little more apt to shoot this issue to Theano then. Don't really see what makes this so complicated. They are already building a  kernel that does a cudaMemcpy(float* dst_block, float* src_block, const uint64_t count, cudaMemcpyHostToDevice) --> run a kernel based on available thread/block counts [i.e. the graph structure that they generate] --> do reduction --> memcpy DeviceToHost. Once this is compiled w/nvcc it should be available to be called again using the header prototype that it generates. 


Regarding the automatic differentiation: why is this necessary? For the cost functions defined in keras don't we have well defined derivatives anyhow? I.e. hinge-loss/l2/... Would writing these derivatives up manually reduce the compile time?

**Comment by patyork** [comment]
I'm not sure - I know some parts of the compilation process are cached, so many times, subsequent compilations can be faster.

Automatic Differentiation is one of the best things about Theano. As for coding them manually to reduce compile time... I couldn't say for sure. The gradient function takes an argument `known_grads=` that may come in handy. I've been curious about that for some of the more involved loss functions, but never so curious as to do it.

**Comment by jramapuram** [comment]
Response from the Theano folks:

		compiled c code is already cached. what take times is probably the optimization time. This is already on our todo list to optimize this.

		You can pickle/unpickle the Theano function to save on the compile time. You must also use this Theano flags: http://deeplearning.net/software/theano/library/config.html#config.reoptimize_unpickled_function

		to False.

		If that don't do it for now while we continue to work on this, tell us.

**Comment by jramapuram** [comment]
@fchollet : I'm guessing that model saving (not model weights) is being tracked somewhere else?

**Comment by fchollet** [comment]
@jramapuram what do you mean by model saving? So far you can only save the weights of a Keras model (for efficiency reasons).

**Comment by jramapuram** [comment]
I am referring to the actual function itself (i.e. the graph that theano builds). 
It takes quite a while to build a graph for some of these recursive structures & if re-used a lot causes quite a lot of wait time unnecessarily.

**Comment by fchollet** [comment]
You could save the entire graph using cPickle. That itself can take a while though.

**Comment by wac81** [comment]
@fchollet  how to do using cPickle?
i try to do save model with 'cPickle.dump(model,open('model.pkl','wb'),pickle.HIGHEST_PROTOCOL)' get a  error

**Comment by yetilihu** [comment]
how can i reuse the model without compile

**Comment by yetilihu** [comment]
?

**Comment by nouiz** [comment]
Update Theano to the development (github) version. There is many updates
related to scan, including compilation time. I think keras implementation
use it.

**Comment by earonesty** [comment]
Unfortunately, for me, even in the latest theano, loading a compiled and pickled model... triggers a compile anyway.

**Comment by philipperemy** [comment]
Anyone did a comparison with TF?

**Comment by lukovkin** [comment]
@philipperemy Hadn't measured it exactly, but TF compiles times faster. May be 10x faster or something close to it.

**Comment by nouiz** [comment]
I don't know what was mean by latest Theano, was it the last release of the
development version. We released Theano 0.8 since that email. It compile
much faster.

Also, the first compilation is slower the then others, as we compile C
code. We cache it and reuse it, so following compilation are faster.

**Comment by locriti** [comment]
Hi @nouiz,
Sorry to possibly hijack this issue discussion, but I'm also experiencing slow "theano.function" calls when attempting to use LSTMs (some 8+ hours). 
This was observed on both 0.8.2 and 0.9.0dev2.dev-RELEASE

To eliminate slower optimization paths, I've profiled this using the latter, running on a i7-2670QM machine with 6g RAM, and THEANO_FLAGS set to:
                            "device=cpu,\
                            floatX=float32,\
                            mode=FAST_RUN,\
                            optimizer_excluding=inplace,\
                            optimizer_including=local_remove_all_assert,\
                            profile=True,\
                            profile_optimizer=True,\
                            profile_memory=True"
Switching to FAST_COMPILE, and targeting "gpu" had no effect.

Attached are the profiling outputs using a [vector](https://github.com/fchollet/keras/files/401208/vector_target_err.txt) target, with which theano.function took 50848.258s, and a [matrix](https://github.com/fchollet/keras/files/401209/matrix_target_err.txt) one taking 52382.505s. These (absolute) times are about 30% longer than what I usually observe due to having two other similar programs running (targeting the gpu), but I imagine relative timings in profiling output are still indicative.
(Outputs are a bit scrambled due to an unrelated failure while invoking print_summary() making them popup in stderr rather than directed to file, but) I'd really appreciate it if you could take a look and share any insights you might have.

**Comment by philipperemy** [comment]
How big is your LSTM? I've never experienced more than 5min on a core i7 6 cores with 10M+ parameters

**Comment by locriti** [comment]
Sorry, this was taken on model with just over 1M parameters, but I
experienced similar behavior on ones smaller than that.

**Comment by philipperemy** [comment]
Can you delete your theano temp dir? The best would be if you could try to reproduce on another machine, e.g. Google Cloud or AWS.
